File: pypp_cli/test_dir/cpp/src/ref_vars.cpp
#include "ref_vars.h"
#include "py_dict.h"
#include "py_str.h"
#include "pypp_util/print.h"

namespace me {
static pypp::PyList<int> &_get_list(pypp::PyList<int> &input_list) {
    return input_list;
}

pypp::PyList<int> &repeat(pypp::PyList<int> &a) {
    a *= 2;
    return a;
}

pypp::PyList<int> repeat_new(pypp::PyList<int> &a) { return a * 2; }

void ref_vars_fn() {
    pypp::print(pypp::PyStr("REF VARS RESULTS:"));
    pypp::PyList<pypp::PyList<int>> a(
        {pypp::PyList({1, 2}), pypp::PyList({3, 4})});
    pypp::print(a);
    pypp::PyList<int> b = a[0];
    b.append(5);
    pypp::print(pypp::PyStr("Python and C++ should print different results:"));
    pypp::print(a);
    pypp::PyList<pypp::PyList<int>> c(
        {pypp::PyList({1, 2}), pypp::PyList({3, 4})});
    pypp::PyList<int> &d = c[0];
    d.append(5);
    pypp::print(pypp::PyStr("Python and C++ should print the same results:"));
    pypp::print(c);
    pypp::PyDict<int, pypp::PyList<int>> e = {{0, pypp::PyList({1, 2})},
                                              {1, pypp::PyList({3, 4})}};
    pypp::PyList<int> &f = e[0];
    f.append(5);
    pypp::print(e);
    pypp::PyList<int> original_list({1, 2, 3});
    pypp::PyList<int> &g = _get_list(original_list);
    g.append(4);
    pypp::print(original_list);
    pypp::PyList<int> arr2({1, 2});
    pypp::PyList<int> &arr2_ref = repeat(arr2);
    arr2_ref.append(42);
    pypp::print(arr2);
    pypp::PyList<int> arr3({1, 2});
    pypp::PyList<int> arr_3_copy = repeat(arr3);
    arr_3_copy.append(42);
    pypp::print(arr3);
    pypp::PyList<int> arr4({1, 2});
    pypp::PyList<int> arr4_copy = repeat_new(arr4);
    arr4_copy.append(42);
    pypp::print(arr4);
    pypp::PyList<int> arr5({1, 2});
    pypp::PyList<int> &arr5_ref = arr5;
    arr5_ref.append(42);
    pypp::print(arr5);
}

} // namespace me

File: pypp_cli/test_dir/python/src/ref_vars.py
from pypp_python import Ref


def _get_list(input_list: list[int]) -> Ref(list[int]):
    return input_list


def repeat(a: list[int]) -> Ref(list[int]):
    a *= 2
    return a


def repeat_new(a: list[int]) -> list[int]:
    return a * 2


def ref_vars_fn():
    print("REF VARS RESULTS:")
    a: list[list[int]] = [[1, 2], [3, 4]]
    print(a)
    # copy
    b: list[int] = a[0]
    b.append(5)
    print("Python and C++ should print different results:")
    print(a)
    # reference
    c: list[list[int]] = [[1, 2], [3, 4]]
    d: Ref(list[int]) = c[0]
    d.append(5)
    print("Python and C++ should print the same results:")
    print(c)
    # with dicts
    e: dict[int, list[int]] = {0: [1, 2], 1: [3, 4]}
    f: Ref(list[int]) = e[0]
    f.append(5)
    print(e)
    # a function that returns a reference
    original_list: list[int] = [1, 2, 3]
    g: Ref(list[int]) = _get_list(original_list)
    g.append(4)
    print(original_list)

    # repeat
    arr2: list[int] = [1, 2]
    arr2_ref: Ref(list[int]) = repeat(arr2)
    arr2_ref.append(42)
    print(arr2)  # should print [1, 2, 1, 2, 42]
    # recommended to not do this.
    arr3: list[int] = [1, 2]
    arr_3_copy: list[int] = repeat(arr3)
    arr_3_copy.append(42)
    # should print [1, 2, 1, 2] in C++, but [1, 2, 1, 2, 42] in Python!
    print(arr3)

    # repeat new
    arr4: list[int] = [1, 2]
    arr4_copy: list[int] = repeat_new(arr4)
    arr4_copy.append(42)
    print(arr4)  # should print [1, 2] for both C++ and Python

    # just a reference to a list
    arr5: list[int] = [1, 2]
    arr5_ref: Ref(list[int]) = arr5
    arr5_ref.append(42)
    print(arr5)  # should print [1, 2, 42] for both C++ and Python
